overseer/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+		let mut this = Self {
+			subsystems: HashMap::new(),
+			id_to_children: HashMap::new(),
+		};
+
+		for s in subsystems.into_iter() {
+			let _ = this.spawn(s);
+		}
+
+		this
+	}
+
+	/// Run the `Overseer`.
+	// TODO: we have to
+	//   * Give out to the user some handler to communicate with the `Overseer`
+	//     to tell it to do things such as `Start` `Stop` or `Spawn`
+	//   * Actually implement stopping of the `Overseer`, atm it's unstoppable.
+	pub async fn run(mut self) {
+		loop {
+			// Upon iteration of the loop we will be collecting all the messages
+			// that need dispatching (if any).
+			let mut msgs = Vec::default();
+
+			for (id, s) in self.subsystems.iter_mut() {
+				if let Some(s) = &mut s.instance {
+					// TODO: set to `None` or restart.
+					// It will panic when the `Future` becomes
+					// ready and then it will be polled again here.
+					if let Poll::Ready(res) = poll!(&mut s.f.0) {
+						log::info!("subsystem stopped {:?}", res);
+					}
+
+					match poll!(&mut s.rx.next()) {
+						Poll::Ready(Some(msg)) => {
+							log::info!("Received message from subsystem {:?}", msg);
+							msgs.push((*id, msg));
+						}
+						_ => {}
+					}
+				}
+			}
+
+			// Do the message dispatching be it broadcasting or direct messages.
+			//
+			// TODO: this desperately need refactoring.
+			for msg in msgs.into_iter() {
+				match msg.1 {
+					OverseerMessage::SubsystemMessage{ to, msg: m } => {
+						match to {
+							Some(to) => {
+								if let Some(subsystem) = self.subsystems.get_mut(&to) {
+									if let Some(ref mut i) = subsystem.instance {
+										let _ = i.tx.send(m).await;
+									}
+								}
+							}
+							None => {
+								for (id, s) in self.subsystems.iter_mut() {
+									// Don't send messages back to the sender.
+									if msg.0 == *id {
+										continue;
+									}
+
+									if s.subsystem.can_recv_msg(&m) {
+										if let Some(ref mut i) = s.instance {
+											let _ = i.tx.send(m.clone()).await;
+										}
+									}
+								}
+							}
+						}
+					}
+					OverseerMessage::Spawn { s, res } => {
+						log::info!("Spawn message");
+
+						let s = self.spawn(s);
+
+						if let Ok(id) = s {
+							match self.id_to_children.get_mut(&msg.0) {
+								Some(ref mut v) => {
+									v.insert(msg.0);
+								}
+								None => {
+									let mut hs = HashSet::new();
+									hs.insert(id);
+									self.id_to_children.insert(msg.0, hs);
+								}
+							}
+						}
+						let _ = res.send(s);
+					}
+				}
+			}
+
+			// Log our "poor man's" process tree as it is at the moment.
+			log::trace!("Pids {:?}", self.id_to_children);
+
+			// Looks like nothing is left to be polled, let's take a break.
+			pending!();
+		}
+	}
+
+	fn spawn(&mut self, mut s: Box<dyn Subsystem<M>>) -> Result<SubsystemId, SubsystemError> {
+		let (to_tx, to_rx) = mpsc::channel(1024);
+		let (from_tx, from_rx) = mpsc::channel(1024);
```

Comment:
const?

Reply:
Just a thought for the future:

It would be great to have this configurable at some point. Running our tests with a channel capacity of 1 (one element buffered, not a rendezvous-channel) makes deadlock detection easier.